### ISSUE TITLE
Implement a plugin system for additional stores.

### ISSUE THREAD

**Issue by Renari** [Heroic-Games-Launcher/HeroicGamesLauncher#3158]
### Feature description

There are various stores that people would like supported on Heroic, and this list is potentially ever growing. I'm proposing implementing a plugin system for additional stores support.

### Additional information

I think this is the best way forward, it would prevent the codebase from becoming bloated with tons of stores that not every user would want.

**Comment by Renari** [comment]
Relevant issues:
#1223
#2974
#3138
#1096
#3021

**Comment by JakobDev** [comment]
Sounds good. How will this work? Do you simple enter th epath to a custom executable that takes the same Arguments and has the same Output like legendary or gogdl?

**Comment by Renari** [comment]
I was thinking of having a standard interface for plugins and expecting a script to be provided that adheres to that interface, this would allow for more flexibility on the plugin authors side since if there is an existing library for that platform they would be able to translate those calls rather than have for fork that library.

**Comment by imLinguin** [comment]
One idea is to do it similarly to GOG Galaxy with IPC. Like I described here 
https://gist.github.com/imLinguin/9b570476f7c6a80ab1bf9e14321b35fa

**Comment by Renari** [comment]
I've found GOG's approach to be extremely error prone after using it for a few years, this would also add complexity since Heroic would have to monitor those processes. 
- What do you do if a plugin crashes?
- What if it's repeatedly crashing and how would we determine that?
> Heroic will request list of internal app ids from each plugin and obtain common metadata from external endpoint like https://gamesdb.gog.com/ or custom equivelent.
I think this should probably be served by the plugin in some way since the closest we have to an all encompassing database is IGDB, but even there they are missing things, something I've encountered frequently when playing games from itch.

**Comment by imLinguin** [comment]
> I've found GOG's approach to be extremely error prone after using it for a few years

I agree it has some issues, but it's up to the plugin to be stable and handle errors gracefully.  
I think most issues for GOG's approach happen because plugins are maintained only by community (with few exceptions). 

> I think this should be served by the plugin since the closest we have to an all encompassing database is IGDB, but even there they are missing things, something I've encountered frequently when playing games from itch.

Centralized database would be required if we wanted to develop capabilities of de-duplicating library entries and having common tags for more advanced filtering. Even if we don't go with that direction idea of this common metadata service is something we kind of need nevertheless.

This is purely an idea and a matter of discussion. Other ideas are welcome

**Comment by Renari** [comment]
> Centralized database would be required if we wanted to develop capabilities of de-duplicating library entries and having common tags for more advanced filtering. Even if we don't go with that direction idea of this common metadata service is something we kind of need nevertheless.

I believe this to be a separate issue, since how would said plugins obtain this id? The store itself wouldn't have references to a central database so it would make sense to do this further up the chain.

**Comment by imLinguin** [comment]
The plugins wouldn't need that info, Heroic could handle it internally, plugins would only need to care about their service's data

**Comment by Renari** [comment]
I foresee syncing with a database, at least as part of the plugin system being a massive burden on this being implemented, it seems like it would better be left as an entirely different feature. Since you have lots of edge cases when trying to do that.

The first example that comes to mind is if someone were to make an emulator plugin, they have Final Fantasy III in their library, you identify it as Final Fantasy III, but it turns out it was the NA version of Final Fantasy III which is actually Final Fantasy VI. Now you need features for the user to split entries and manually match them against a database.

**Comment by Renari** [comment]
@imLinguin I thought about this recently and Plex, Jellyfin etc already do something similar to what you were proposing via plugins that act as metadata providers. This wouldn't work exactly like you describe in your post, you would need to use internal ids instead of gamedb ids and then have a table to map them based on metadata providers. This would cover all cases because if there's a game that's missing metadata it just means a meatadata provider that supports that use-case hasn't been made yet.

**Comment by cerels** [comment]
Bump, this feature could put heroic on par with playnite while also lifting work from the devs

**Comment by TopHatTurtle97** [comment]
Also bump, heroic is hands down the cleanest solution for getting non steam games on steam OS, a plugin store system would be beneficial to both the developers long term and the users.

**Comment by poyo86** [comment]
Bump, this would be awesome

**Comment by MatteoAlberghini** [comment]
Good usecase is something like https://romm.app/, they already mantain a playnite plugin and having this library integrate and runnable on linux would be awesome.
